src/natives/java_util.ts

```diff
@@ -307,35 +307,131 @@ class java_util_zip_CRC32 {
 
 class java_util_zip_Deflater {
 
-  public static 'initIDs()V'(thread: JVMThread): void {
-    thread.throwNewException('Ljava/lang/UnsatisfiedLinkError;', 'Native method not implemented.');
-  }
+  public static 'initIDs()V'(thread: JVMThread): void {}
 
-  public static 'init(IIZ)J'(thread: JVMThread, arg0: number, arg1: number, arg2: number): Long {
-    thread.throwNewException('Ljava/lang/UnsatisfiedLinkError;', 'Native method not implemented.');
-    return null;
+  /**
+   * Initialize a new deflater. Using the zlib recommended default values.
+   */
+
+  public static 'init(IIZ)J'(thread: JVMThread, level: number, strategy: number, nowrap: number): Long {
+    let DEF_MEM_LEVEL = 8; // Zlib recommended default
+    let Z_DEFLATED = 8;    // This value is in the js version of pako under pako.Z_DEFLATED. 
+    // Possibly it is set to private in the Typescript version. The default value is 8, so this should work fine
+
+    let strm = new ZStreamCons();
+    let ret = deflate.deflateInit2(strm, level, Z_DEFLATED, nowrap ? -MAX_WBITS : MAX_WBITS, DEF_MEM_LEVEL, strategy);
+    
+    if (ret != ZlibReturnCode.Z_OK) {
+      let msg = ((strm.msg != null) ? strm.msg :
+                (ret == ZlibReturnCode.Z_STREAM_ERROR) ?
+                "inflateInit2 returned Z_STREAM_ERROR" :
+                "unknown error initializing zlib library");
+      thread.throwNewException("Ljava/lang/InternalError;", msg);
+    }
+    let num = OpenZStream(strm);
+    return Long.fromNumber(num);
   }
 
+  /**
+   * Apparently this is explicitly not supported by pako.
+   * @see Notes at http://nodeca.github.io/pako/
+   */
+
   public static 'setDictionary(J[BII)V'(thread: JVMThread, arg0: Long, arg1: JVMTypes.JVMArray<number>, arg2: number, arg3: number): void {
     thread.throwNewException('Ljava/lang/UnsatisfiedLinkError;', 'Native method not implemented.');
   }
 
-  public static 'deflateBytes(J[BIII)I'(thread: JVMThread, javaThis: JVMTypes.java_util_zip_Deflater, arg0: Long, arg1: JVMTypes.JVMArray<number>, arg2: number, arg3: number, arg4: number): number {
-    thread.throwNewException('Ljava/lang/UnsatisfiedLinkError;', 'Native method not implemented.');
+  public static 'deflateBytes(J[BIII)I'(thread: JVMThread, javaThis: JVMTypes.java_util_zip_Deflater, addr: Long, b: JVMTypes.JVMArray<number>, off: number, len: number, flush: number): number {
+    let strm = GetZStream(thread, addr.toNumber());
+    if (!strm) return;
+
+    let thisBuf = javaThis['java/util/zip/Deflater/buf'];
+    let thisOff = javaThis['java/util/zip/Deflater/off'];
+    let thisLen = javaThis['java/util/zip/Deflater/len'];
+
+    let res: Number;
```

Comment:
This should probably be `number`? Hmm, unless TypeScript also accepts `Number! Primitives are typically lowercase.

You can actually move this declaration to where you assign it. `let` is block-scoped, so you can declare it in the `if` and the `else` branches without conflict. Then, you don't even need to annotate it.